### PR TITLE
Update tangle.py

### PR DESCRIPTION
**Pull request by Vannghia69**
The updated file with a new untangle method will solve the tanglegram layout problem more effectively than the random method

### CODE REVIEW

File: tanglegram/tangle.py [hunk 1]
```diff
@@ -171,73 +167,148 @@ def gen_tangle(a, b, labelsA=None, labelsB=None, optimize_order=10000,
...
+            for j in all_linkage2:
+                best_linkage1 = i
+                best_linkage2 = j
+                dend1 = sclust.hierarchy.dendrogram(best_linkage1, labels=labels1)
+                dend2 = sclust.hierarchy.dendrogram(best_linkage2, labels=labels2)
+                min_entang = get_entanglement(dend1, dend2)
+                old_rem = -1
+                if len(save_entang) == 0:
+                    save_entang.append(min_entang)
+                for k in range(10):
```
Comment (schlegelp) [inline tanglegram/tangle.py:284]: Just having a quick glance - will have a thorough look later but where is this hardcoded `range(10)` coming from?

File: tanglegram/tangle.py [hunk 2]
```diff
@@ -171,73 +167,148 @@ def gen_tangle(a, b, labelsA=None, labelsB=None, optimize_order=10000,
...
+                    final_linkage2 = best_linkage2
+                    final_entang = min_entang
+                save_entang.append(final_entang)
+                
+        if old_entang == final_entang:
+            cond += 1
+        old_entang = final_entang
+        if cond == 2:
+            break
+    plt.clf()
```
Comment (schlegelp) [inline tanglegram/tangle.py:309]: I'm not sure this is necessary here?

File: tanglegram/tangle.py [hunk 3]
```diff
@@ -171,73 +167,148 @@ def gen_tangle(a, b, labelsA=None, labelsB=None, optimize_order=10000,
...
+
+    for li_MID in range(len(linkage_matrix1) - 1, 1, -1):
+        all_linkage1 = get_all_linkage(linkage_matrix1, li_MID)
+        all_linkage2 = get_all_linkage(linkage_matrix2, li_MID)
+
+        for i in all_linkage1:
+            for j in all_linkage2:
+                best_linkage1 = i
+                best_linkage2 = j
+                dend1 = sclust.hierarchy.dendrogram(best_linkage1, labels=labels1)
```
Comment (schlegelp) [inline tanglegram/tangle.py:278]: use `no_plot=True` to avoid plotting the dendrograms

**Comment by Vannghia69** [comment]
Hi,


  *   I made the range(10) as a convergence condition. It stops when the result repeats twice (cannot be improved)
  *   The "Fine optimization" is necessary because it helped to reduce the entanglement value more. For example from 1.1 to 1.0. I have seen its effectiveness.
  *   Yes, I should use no_plot = True. I tried to prevent it from making plots. This is really useful.

________________________________

File: tanglegram/tangle.py (hunk 1 above)
Comment (Vannghia69) [reply tanglegram/tangle.py:284]: I can change the convergence condition to make it more sensible

**Comment by schlegelp** [comment]
I'll have to have a thorough look at your proposed changes. In the meantime, could you perhaps provide an example comparing the existing and your new method to showcase that it is either faster and/or produces better results?

**Comment by Vannghia69** [comment]
Yes. I will provide you with an example of how the improvement looks like.
________________________________

**Comment by Vannghia69** [comment]
Here is an example which is the result of a project that I am working with.

From the left to the right: Original tanglegram (Entanglement value = 6.0), Tanglegram with random untangle method ( Entanglement value = 0.6), and Tanglegram with new untangle method (Entanglement value = 0.2).

[cid:d8b64736-e277-4788-bcf4-82cbe4434964][cid:63618ed9-92a0-4383-8436-8fb286872d1f][cid:c17b174a-786c-4a85-9bda-6439391cacd0]
________________________________

**Comment by Vannghia69** [comment]
In the tanglegram optimized by the random method, you can see that there are some lines that intersecting each other but we can actually switch objects positions to reduce the entanglement value. For example, 2 objects WcembelSS and MudfineSS. All I want to do is finding the global minimum of entanglement values, not the local minimums.
________________________________

**Comment by schlegelp** [comment]
Unfortunately, embedding images doesn't appear to work if you respond to the email. Could you go the Github pull request and post them there?

**Comment by Vannghia69** [comment]
Sure!
________________________________

**Comment by Vannghia69** [comment]
23random06
23nghia02

**Pull request merged**